feat(oxc_transformer): replace_global_define ThisExpression

IWANABETHATGUY opened the pull request:
Test from https://github.com/evanw/esbuild/blob/d34e79e2a998c21bb71d57b92b0017ca11756912/internal/bundler_tests/bundler_default_test.go#L5195-L5260, 

esbuild snapshot: 

https://github.com/evanw/esbuild/blob/d34e79e2a998c21bb71d57b92b0017ca11756912/internal/bundler_tests/snapshots/snapshots_default.txt#L1081-L1108

IWANABETHATGUY commented:
* **#7443** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7443?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7443?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented on crates/oxc_transformer/src/plugins/replace_global_defines.rs:
```diff
@@ -18,9 +19,19 @@ use oxc_traverse::{traverse_mut, Traverse, TraverseCtx};
 pub struct ReplaceGlobalDefinesConfig(Arc<ReplaceGlobalDefinesConfigImpl>);
 
+lazy_static! {
+    static ref THIS_ATOM: Atom<'static> = Atom::from("this");
```
We probably don't need this.

Boshen commented on crates/oxc_transformer/src/plugins/replace_global_defines.rs:
```diff
@@ -446,12 +476,16 @@ impl<'a> ReplaceGlobalDefines<'a> {
                         None
                     }
+                    Expression::ThisExpression(_) if should_replace_this_expr => {
+                        cur_part_name = &THIS_ATOM;
```
`cur_part_name.as_str() == "this"` should do the same thing.

IWANABETHATGUY replied:
Just want to make the code logic consistent, if use `cur_part_name.as_str() == "this"`, you need to check if the cursor is zero.

Boshen commented:
@IWANABETHATGUY I merged in the wrong order, can you rebase?

IWANABETHATGUY commented:
> @IWANABETHATGUY I merged in the wrong order, can you rebase?

ok

Boshen commented:
### Merge activity

* **Nov 24, 11:13 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

IWANABETHATGUY committed:
feat(oxc_transformer): replace_global_define ThisExpression (#7443)

Test from https://github.com/evanw/esbuild/blob/d34e79e2a998c21bb71d57b92b0017ca11756912/internal/bundler_tests/bundler_default_test.go#L5195-L5260,

esbuild snapshot:

https://github.com/evanw/esbuild/blob/d34e79e2a998c21bb71d57b92b0017ca11756912/internal/bundler_tests/snapshots/snapshots_default.txt#L1081-L1108
